environment variable to show debug options

Erotemic opened the pull request:
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.


Digging through the code I saw there were a lot of builtin tools for profiling / debugging, but (as far as I can tell) there was no mention of these in the docs, nor was there wany way for a user to become aware of them without digging into the code.

To remedy this, I added an environment variable `DVC_SHOW_DEBUG_OPTIONS` (which perhaps could be shorted to `DVC_DEBUG`), and if this is specified the hidden CLI options will be shown to the user. Otherwise the behavior is the same. 

Also, when I tried to run a command with `--show-stack` I didn't see any difference in the output. So I'm not sure what it's actually doing.

I've added a section in the contributing docs which mentions this new option. The corresponding PR is here: https://github.com/iterative/dvc.org/pull/4662

skshetry commented:
They are documented in wiki, https://github.com/iterative/dvc/wiki/Debugging,-Profiling-and-Benchmarking-DVC, although they may be hard to find.


> Also, when I tried to run a command with `--show-stack` I didn't see any difference in the output. So I'm not sure what it's actually doing.

It only works on macOS and Linux. On macOS, you have to press `Ctrl + T` and on Linux, you have to press `Ctrl + \ (backslash)`.

Erotemic commented:
Thanks for the pointer! It probably makes sense to replace the blurb I wrote in the contributing docs with a reference to this wiki. I think having this info in contributing docs will make it a lot easier to find because the people who would need debugging tools (i.e. contributors) should be reading that.

Erotemic commented:
I updated the help strings based on info in the wiki, and I also modified the docs PR to point to the wiki.

efiop commented on dvc/_debug.py:
```diff
@@ -196,35 +196,83 @@ def debugtools(args: Optional["Namespace"] = None, **kwargs):
-    parser.add_argument("--yappi", action="store_true", default=False, help=SUPPRESS)
+    # For detailed info see:
+    # https://github.com/iterative/dvc/wiki/Debugging,-Profiling-and-Benchmarking-DVC
+    dvc_show_debug_options = os.environ.get("DVC_SHOW_DEBUG_OPTIONS", "")
```
Seems a bit too verbose. Maybe better to use `--debug` flag? Do you know maybe some examples of how other projects handle this?

skshetry replied:
In the long term, I'd really like to introduce the concept of "extensions", even if it's for our own use. We could enable extra functionalities if the `devel` extension is enabled, for example, in the config.

Erotemic replied:
I tend to opt for verbosity when I'm unsure. I think `DVC_DEBUG=1 <cmd>` would likely work as well.

I don't see an easy way to add a `--debug` option because the argparse object hasn't finished being constructed at the time where we need to determine if we use SUPRESS or add a help message. I think the environment variable approach is the cleanest way to allow these options to be conditionally exposed in `--help`.

The other options I can think of are:

* Add a `--debug` flag which prints a special help section, but that would need to be maintained separately, so that's undesirable. 

* Check if `"--debug" in sys.argv`, but that's undesirable because 1. the parser might not be applied to sys.argv, and 2. there is a chance `--debug` might appear on the command line, but not be intended as an argument flag (e.g. `dvc add -- --debug`). Checking `sys.argv` won't respect the `--` like argparse will. These issues probably won't arise in practice, and even if they did it would likely not matter because it just changes the content of the help message. Still, I think it is unclean design and the environment variable approach seems like the "idomatic" thing to do here. That being said,  `"--debug" in sys.argv` is simple, and I'm open to doing it this way. The only thing I think is important is that there is a way for a developer to unhide these options.

efiop replied:
@skshetry Seems like a total overkill for this purpose. But I'm not even sure what extensions you mean, I suppose like git extensions? Don't quite see the need.

efiop replied:
@Erotemic DVC_DEBUG sounds fine to me, let's not overthink it.

efiop replied:
@Erotemic Also note dvc/env.py where we try to keep all env vars in.

efiop replied:
@Erotemic regarding the cli option, there is a way to partially parse known options in argparse IIRC. Just checking argv (not sys.argv) is also fine.

Erotemic replied:
Switched to DVC_DEBUG and added it to env.

I do like the idea of showing the extra help with `vebose` is set. At first I wasn't sure how to do it, but I think its possible as long as the parser is initialized with `add_help=False` (which it is) and if we can ensure the verbose argument is added to the parser before we call `add_debugging_flags`, which should just require reordering the code in `get_parent_parser`.

Now that I see how to do this, I think it would be a good idea to remove the DVC_DEBUG environment variable entirely, and just add the extra help options is `-v` is already given (which will make it much easier for a developer to stumble on this too, which is good).

Erotemic replied:
I made the above change, I can revert if you want to go with the environ, but I think just respecting `-v` is much cleaner now that I know about the `add_help` argment to `ArgumentParser`.

efiop replied:
Thank you @Erotemic and @skshetry 🙏

Erotemic commented:
While playing with this I noticed two things and made changes accordingly. First the "cprofile-dump" option was not near "cprofile", so I moved it to keep similar options grouped. Secondly, if you run `DVC_SHOW_DEBUG_OPTIONS=1 dvc add --help` the debug options were intersperced with the add options, so I added an argument group such that they are always displayed separately. E.g.

```
(pyenv3.11.2) joncrall@toothbrush:~/remote/toothbrush/data/dvc-repos/smart_data_dvc-ssd/Drop7-Cropped2GSD/CH_R001/CH_R001_CLUSTER_041$ DVC_SHOW_DEBUG_OPTIONS=1 dvc add --help
usage: dvc add [-h] [--cprofile] [--cprofile-dump CPROFILE_DUMP] [--yappi] [--yappi-separate-threads] [--viztracer] [--viztracer-depth VIZTRACER_DEPTH] [--viztracer-async] [--pdb]
               [--instrument] [--instrument-open] [--show-stack] [-q | -v] [--no-commit] [--glob] [-o <path>] [--to-remote] [-r <name>] [--remote-jobs <number>] [-f]
               targets [targets ...]

Track data files or directories with DVC.
Documentation: <https://man.dvc.org/add>

positional arguments:
  targets               Input files/directories to add.

options:
  -h, --help            show this help message and exit
  -q, --quiet           Be quiet.
  -v, --verbose         Be verbose.
  --no-commit           Don't put files/directories into cache.
  --glob                Allows targets containing shell-style wildcards.
  -o <path>, --out <path>
                        Destination path to put files to.
  --to-remote           Download it directly to the remote
  -r <name>, --remote <name>
                        Remote storage to download to
  --remote-jobs <number>
                        Only used along with '--to-remote'. Number of jobs to run simultaneously when pushing data to remote.The default value is 4 * cpu_count().
  -f, --force           Override local file or folder if exists.

debug options:
  --cprofile            Generate cprofile data for tools like snakeviz / tuna
  --cprofile-dump CPROFILE_DUMP
                        Location to dump cprofile file
  --yappi               Generate a callgrind file for use with tools like kcachegrind / qcachegrind
  --yappi-separate-threads
                        Generate one callgrind file per thread
  --viztracer           Generate a viztracer file for use with vizviewer
  --viztracer-depth VIZTRACER_DEPTH
                        Set viztracer maximum stack depth
  --viztracer-async     Treat async tasks as threads
  --pdb                 Drop into the pdb/ipdb debugger on any exception
  --instrument          Use pyinstrument CLI profiler
  --instrument-open     Use pyinstrument web profiler
  --show-stack, --ss    Use Ctrl+T on macOS or Ctrl+\ on Linux to print the stack frame currently executing. Unavailable on Windows.
```

skshetry commented on dvc/_debug.py:
```diff
@@ -198,33 +198,83 @@ def debugtools(args: Optional["Namespace"] = None, **kwargs):
-    parser.add_argument("--yappi", action="store_true", default=False, help=SUPPRESS)
+    # For detailed info see:
+    # https://github.com/iterative/dvc/wiki/Debugging,-Profiling-and-Benchmarking-DVC
+    dvc_show_debug_options = parser.parse_known_args()[0].verbose
```
Nitpicky suggestion, I think it's more clear that we are using `verbose` as a condition for showing the message.

```suggestion
    args, _ = parser.parse_known_args()
    verbose = args.verbose
```
